File: src/my_cpp_pkg/src/number_counter.cpp
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

#include "rclcpp/rclcpp.hpp"

class NumberCounter : public rclcpp::Node {
public:
  NumberCounter() : rclcpp::Node("number_counter") {
    RCLCPP_INFO(this->get_logger(), "Number Counter CPP Node has Strated!");
    this->num_count_publisher_ =
        this->create_publisher<example_interfaces::msg::Int64>("number_count",
                                                               10);
    this->num_reciever_ =
        this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10, [this](example_interfaces::msg::Int64 msg) {
              this->numRecCallback(msg);
            });
    this->counter_reset_service_ =
        this->create_service<example_interfaces::srv::SetBool>(
            "reset_counter",
            [this](example_interfaces::srv::SetBool::Request::SharedPtr req,
                   example_interfaces::srv::SetBool::Response::SharedPtr res) {
              this->resetCounter(req, res);
            });
  }

private:
  rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr
      num_count_publisher_;
  rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr num_reciever_;
  int count_{0};
  rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr
      counter_reset_service_;

  void numRecCallback(example_interfaces::msg::Int64 msg) {
    RCLCPP_INFO(this->get_logger(), "Number Recd: %d", (int)msg.data);
    this->count_ += 1;
    this->publishNumberCount();
  }

  void publishNumberCount() {
    auto msg = example_interfaces::msg::Int64();
    msg.data = count_;
    this->num_count_publisher_->publish(msg);
    RCLCPP_INFO(this->get_logger(), "Number Count Published");
  }

  void
  resetCounter(const example_interfaces::srv::SetBool::Request::SharedPtr req,
               example_interfaces::srv::SetBool::Response::SharedPtr res) {

    if (req->data) {
      this->count_ = 0;
      res->success = true;
      res->message = "Counter Reset Successful";
    }
  }
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<NumberCounter>();
  rclcpp::spin(node);
  rclcpp::shutdown();
}

File: src/my_py_pkg/my_py_pkg/number_counter.py
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool


class NumberCounter(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.num_count_ = 0
        self.num_count_publisher_ = self.create_publisher(Int64, "number_count", 10)
        self.num_reciever = self.create_subscription(
            Int64, "number", self.numRecCallback, 10
        )
        self.counter_reset_service_ = self.create_service(
            SetBool, "reset_counter", self.resetCounter
        )
        self.get_logger().info("Number Counter Node Started!")

    def publishNumberCount(self):
        msg = Int64()
        msg.data = self.num_count_
        self.num_count_publisher_.publish(msg)
        self.get_logger().info("Number Count Published")

    def numRecCallback(self, msg: Int64):
        self.get_logger().info(f"Number Recd: {msg.data}")
        self.num_count_ += 1
        self.publishNumberCount()

    def resetCounter(self, request: SetBool.Request, response: SetBool.Response):
        if request.data:
            self.num_count_ = 0
            response.success = True
            response.message = "Counter Reset Successfully"
        return response


def main(args=None):
    rclpy.init(args=args)
    node = NumberCounter()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
